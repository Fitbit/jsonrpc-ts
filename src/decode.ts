// tslint:disable-next-line:import-name
import ErrorSubclass from 'error-subclass';
import * as t from 'io-ts';
import { failure } from 'io-ts/lib/PathReporter';

import { Peer } from './peer';

export class DecodeError extends ErrorSubclass {
  static displayName = 'DecodeError';
}

/**
 * Creates a decoder function for an io-ts type.
 *
 * The decoder function throws DecodeError if the input value is not of
 * the expected type.
 *
 * Decoder functions can be used to build wrappers around RPC method
 * calls which enforce type-safety of the return values.
 *
 * ```typescript
 * const myMethod = (args: MyMethodArgs) =>
 *   rpc.callMethod('myMethod', args).then(decode(MyMethodResult));
 * // Return type of myMethod is Promise<TypeOf<MyMethodResult>>
 * ```
 *
 * @param type runtime type to decode
 */
export function decode<A, O, I>(type: t.Type<A, O, I>) {
  return (value: I): A => {
    return type.decode(value).getOrElseL((errors) => {
      throw new DecodeError(failure(errors).join('\n'));
    });
  };
}
