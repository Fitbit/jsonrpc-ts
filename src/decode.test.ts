import * as t from 'io-ts';

import { decode, DecodeError } from './decode';

// Lifted from the io-ts README
// tslint:disable-next-line:variable-name
const DateFromString = new t.Type<Date, string>(
  'DateFromString',
  (m): m is Date => m instanceof Date,
  (m, c) =>
    t.string.validate(m, c).chain((s) => {
      const d = new Date(s);
      return isNaN(d.getTime()) ? t.failure(s, c) : t.success(d);
    }),
  a => a.toISOString(),
);

// it returns a function of the right type
const decodeDate: (value: t.mixed) => Date = decode(DateFromString);

it('returns the decoded value', () => {
  const expectedDate = new Date(2018, 2, 16);
  expect(decodeDate(expectedDate.toISOString())).toEqual(expectedDate);
});

it('throws when decode fails', () => {
  expect(() => decodeDate('foo')).toThrowErrorMatchingSnapshot();
});
