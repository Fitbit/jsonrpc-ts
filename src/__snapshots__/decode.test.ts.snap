// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws when decode fails 1`] = `"Invalid value \\"foo\\" supplied to : DateFromString"`;
